signals/src/signal.rs

```diff
@@ -450,6 +458,109 @@ impl<A> Signal for Flatten<A>
     }
 }
 
+#[doc(hidden)]
+pub struct Broadcast<A: Signal>
```

Comment:
I feel like it should be possible to make this struct have a single-field (which is nice since single-field structs are zero-cost).

Reply:
There's something I must have missed. Why's the single-field case zero-cost compared to the two-field case?

Reply:
I'm not sure if it's guaranteed by the Rust compiler, but generally speaking if you have a struct that has a single field then at runtime it will compile to the same code as if you weren't using the struct at all.

In other words, this:

```rust
struct Foo {
  value: Option<u32>,
}
```

Should be the same at runtime as simply using `Option<u32>` directly. In other words it has the same memory layout as `Option<u32>`

But if you use two or more fields, then Rust has to actually create a two-field struct at runtime (it can't just use the same memory layout).

I'm not sure if it actually makes much difference to performance, it might not make any difference, since structs are very tightly packed and have very little overhead.

Reply:
Ah, nice explanation, thanks! To get a single-field struct in this case it'll probably have to be an Rc or a Box putting everything onto the heap instead. I guess it depends what I can cook up!

Reply:
I was thinking something like this:

```rust
struct BroadcastSharedState<A> {
    signal: A,
    value: Option<A::Item>,
    children: Vec<Weak<RefCell<BroadcastState<A>>>>,
}

struct BroadcastState<A> {
    has_changed: bool,
    task: Task,
    shared_state: Rc<RefCell<BroadcastSharedState<A>>>,
}

pub struct Broadcast<A: Signal> {
    inner: Rc<RefCell<BroadcastState<A>>>,
}
```

Reply:
This might be faster, though:

```rust
struct BroadcastSharedState<A> {
    signal: A,
    value: Option<A::Item>,
    children: Vec<Weak<BroadcastState<A>>>,
}

struct BroadcastState<A> {
    has_changed: Cell<bool>,
    task: RefCell<Task>,
    shared_state: Rc<RefCell<BroadcastSharedState<A>>>,
}

pub struct Broadcast<A: Signal> {
    inner: Rc<BroadcastState<A>>,
}
```

Reply:
Thanks, I was thinking along similar lines, good to see that's what you'd be expecting! I quite like the latter form.